jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/NativeLibLoader.java

```diff
@@ -34,65 +34,127 @@
 class NativeLibLoader {
   private static final Log logger = LogFactory.getLog(NativeLibLoader.class);
 
+  /**
+   * Supported OS enum.
+   */
+  enum OS {
+    WINDOWS("windows"),
+    MACOS("macos"),
+    LINUX("linux"),
+    SOLARIS("solaris");
+
+    final String name;
+
+    private OS(String name) {
+      this.name = name;
+    }
+
+    /**
+     * Detects the OS using the system properties.
+     * Throws IllegalStateException if the OS is not recognized.
+     * @return The OS.
+     */
+    static OS detectOS() {
+      String os = System.getProperty("os.name", "generic").toLowerCase(Locale.ENGLISH);
+      if (os.contains("mac") || os.contains("darwin")) {
+        return MACOS;
+      } else if (os.contains("win")) {
+        return WINDOWS;
+      } else if (os.contains("nux")) {
+        return LINUX;
+      } else if (os.contains("sunos")) {
```

Comment:
Not sure if the BSD family should be added here.  They do package the Python library.

Reply:
It should be straightforward to add. Are there any other platforms? I didn't bother with things like AIX or POWER, but the JVM will run on them.

Reply:
Nah, let's do it when the case arises.  We don't test it on *BSD anyway.